File: Queues & Stack/next_greater_element.cpp
//Problem Link : https://leetcode.com/problems/next-greater-element-i/description/
//Solution Link:https://youtu.be/V5r7PQhcJCQ?si=yoUawrOUbBs5AuVJ
//Complexity: O(N)

class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
    vector <int> nge;
    stack<int> st;
    unordered_map<int,int>map;
    for (int i = 0; i < nums2.size(); i++)
    {
        while((! st.empty())  && (nums2[i]>nums2[st.top()])){
            map[nums2[st.top()]] = nums2[i];
            st.pop();
        }
        st.push(i);
    }
    while(! st.empty()){
        map[nums2[st.top()]] = -1;
        st.pop();
    }
    for(int i=0;i<nums1.size();i++){
        nge.push_back(map[nums1[i]]);
    }
    return nge;   
    }
};

File: Queues & Stack/next-greater-element.py
class Solution(object):
    def nextGreaterElement(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: List[int]
        """

        ans = []

        for num in nums1:
            found = False
            for i in range(len(nums2)):
                if found and num < nums2[i]:
                    ans.append(nums2[i])
                    break
                if nums2[i]==num:
                    found= True
            else:
                ans.append(-1)

        return ans
